File: DentyEngine/src/System/Input/Input.cpp
#include "stdafx.h"

#include "Input.h"

#include "System/Core/DString.h"

namespace DentyEngine
{
	//
	// Static variables.
	//
	Mouse Input::_mouse = { };
	Keyboard Input::_keyboard = { };
	std::array<GamePad, Input::NUM_GAME_PAD> Input::_gamePads = { };

	//
	// Static functions.
	//
	void Input::Update()
	{
		_mouse.Update();
		_keyboard.Update();

		for (auto&& gamePad : _gamePads)
		{
			gamePad.Update();
		}
	}

	void Input::OnGui()
	{
		if (ImGui::CollapsingHeader("Input##CoreLayer"))
		{
			if (ImGui::TreeNodeEx("Mouse##Input"))
			{
				_mouse.OnGui();

				ImGui::TreePop();
			}

			if (ImGui::TreeNodeEx("Keyboard##Input"))
			{
				_keyboard.OnGui();

				ImGui::TreePop();
			}

			if (ImGui::TreeNodeEx("GamePad##Input"))
			{
				_gamePads.at(0).OnGui();

				ImGui::TreePop();
			}
		}
	}

	void Input::SetMousePositionX(int32_t x)
	{
		_mouse.SetMousePositionX(x);
	}

	void Input::SetMousePositionY(int32_t y)
	{
		_mouse.SetMousePositionY(y);
	}

	void Input::SetMousePosition(const Vector2I& position)
	{
		_mouse.SetMousePosition(position);
	}

	void Input::SetMouseWheelSpeed(float wheelSpeed)
	{
		_mouse.SetWheelSpeed(wheelSpeed);
	}

	void Input::ResetMouseWheelSpeed()
	{
		_mouse.ResetWheelSpeed();
	}

	float Input::GetMouseWheelSpeed()
	{
		return _mouse.GetWheelSpeed();
	}

	float Input::GetMousePanX()
	{
		return _mouse.GetPanX();
	}

	float Input::GetMousePanY()
	{
		return _mouse.GetPanY();
	}

	Vector2I Input::GetMousePosition()
	{
		return _mouse.GetMousePosition();
	}

	bool Input::IsMousePressed(MouseButton button)
	{
		return _mouse.IsPressed(button);
	}

	bool Input::IsMouseTriggered(MouseButton button)
	{
		return _mouse.IsTrigger(button);
	}

	bool Input::IsMouseReleased(MouseButton button)
	{
		return _mouse.IsReleased(button);
	}

	bool Input::IsKeyPressed(KeyCode keyCode)
	{
		return _keyboard.IsPressed(keyCode);
	}

	bool Input::IsKeyTriggered(KeyCode keyCode)
	{
		return _keyboard.IsTrigger(keyCode);
	}

	bool Input::IsKeyReleased(KeyCode keyCode) 
	{
		return _keyboard.IsReleased(keyCode);
	}

	uint8_t Input::GetLeftTrigger(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetLeftTrigger();
	}

	uint8_t Input::GetRightTrigger(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetRightTrigger();
	}

	int16_t Input::GetThumbLX(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetThumbLX();
	}

	int16_t Input::GetThumbLY(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetThumbLY();
	}

	int16_t Input::GetThumbRX(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetThumbRX();
	}

	int16_t Input::GetThumbRY(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetThumbRY();
	}

	float Input::GetNormalizedThumbLX(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetNormalizedThumbLX();
	}

	float Input::GetNormalizedThumbLY(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetNormalizedThumbLY();
	}

	float Input::GetNormalizedThumbRX(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetNormalizedThumbRX();
	}

	float Input::GetNormalizedThumbRY(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).GetNormalizedThumbRY();
	}

	bool Input::IsPadPressed(PadKeyCode keyCode, uint32_t padNum) 
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).IsPressed(keyCode);
	}

	bool Input::IsPadTriggered(PadKeyCode keyCode, uint32_t padNum) 
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).IsTrigger(keyCode);
	}

	bool Input::IsPadReleased(PadKeyCode keyCode, uint32_t padNum) 
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).IsReleased(keyCode);
	}

	bool Input::IsPadLeftTriggerTriggered(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).IsLeftTriggerTriggered();
	}

	bool Input::IsPadRightTriggerTriggered(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).IsRightTriggerTriggered();
	}

	bool Input::IsPadConnectedAt(uint32_t padNum)
	{
		DENTY_CHECK_IS_OUT_OF_RANGE_STL(padNum, _gamePads, __func__)

		return _gamePads.at(padNum).Connected();
	}
}

File: DentyEngine-ScriptCore/ScriptCore/Core/Input.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DentyEngine
{
    public static class Input
    {
        //
        // Keyboard
        //
        
        public static bool IsKeyPressed(KeyCode keyCode)
        {
            return InternalCalls.Input_IsKeyPressed(keyCode);
        }

        public static bool IsKeyTriggered(KeyCode keyCode)
        {
            return InternalCalls.Input_IsKeyTriggered(keyCode);
        }

        public static bool IsKeyReleased(KeyCode keyCode)
        {
            return InternalCalls.Input_IsKeyReleased(keyCode);
        }


        //
        // Mouse
        //
        
        public static bool IsMousePressed(MouseButton button)
        {
            return InternalCalls.Input_IsMousePressed(button);
        }

        public static bool IsMouseTriggered(MouseButton button)
        {
            return InternalCalls.Input_IsMouseTriggered(button);
        }

        public static bool IsMouseReleased(MouseButton button)
        {
            return InternalCalls.Input_IsMouseReleased(button);
        }

        public static void SetMousePosition(Vector2I position)
        {
            InternalCalls.Input_SetMousePosition(position.x, position.y);
        }

        public static float GetMouseWheelSpeed()
        {
            float speed = 0.0f;

            InternalCalls.Input_GetMouseWheelSpeed(out speed);

            return speed;
        }

        public static float GetMousePanX()
        {
            float panX = 0.0f;
            
            InternalCalls.Input_GetMousePanX(out panX);

            return panX;
        }

        public static float GetMousePanY()
        {
            float panY = 0.0f;

            InternalCalls.Input_GetMousePanY(out panY);

            return panY;
        }

        public static Vector2I GetMousePosition()
        {
            Vector2I position = new Vector2I();
            
            InternalCalls.Input_GetMousePosition(out position.x, out position.y);

            return position;
        }

        //
        // Gamepad
        //
        public static bool IsPadPressed(PadKeyCode padKeyCode, uint padNum = 0)
        {
            return InternalCalls.Input_IsPadPressed(padKeyCode, padNum);
        }

        public static bool IsPadTriggered(PadKeyCode padKeyCode, uint padNum = 0)
        {
            return InternalCalls.Input_IsPadTriggered(padKeyCode, padNum);
        }

        public static bool IsPadReleased(PadKeyCode padKeyCode, uint padNum = 0)
        {
            return InternalCalls.Input_IsPadReleased(padKeyCode, padNum);
        }

        public static bool IsPadLeftTriggerTriggered(uint padNum = 0)
        {
            return InternalCalls.Input_IsPadLeftTriggerTriggered(padNum);
        }

        public static bool IsPadRightTriggerTriggered(uint padNum = 0)
        {
            return InternalCalls.Input_IsPadRightTriggerTriggered(padNum);
        }

        public static bool IsPadConnectedAt(uint padNum)
        {
            return InternalCalls.Input_IsPadConnectedAt(padNum);
        }

        public static byte GetLeftTrigger(uint padNum = 0)
        {
            return InternalCalls.Input_GetLeftTrigger(padNum);
        }

        public static byte GetRightTrigger(uint padNum = 0)
        {
            return InternalCalls.Input_GetRightTrigger(padNum);
        }

        public static short GetThumbLX(uint padNum = 0)
        {
            return InternalCalls.Input_GetThumbLX(padNum);
        }

        public static short GetThumbLY(uint padNum = 0)
        {
            return InternalCalls.Input_GetThumbLY(padNum);
        }

        public static short GetThumbRX(uint padNum = 0)
        {
            return InternalCalls.Input_GetThumbRX(padNum);
        }

        public static short GetThumbRY(uint padNum = 0)
        {
            return InternalCalls.Input_GetThumbRY(padNum);
        }

        public static float GetNormalizedThumbLX(uint padNum = 0)
        {
            return InternalCalls.Input_GetNormalizedThumbLX(padNum);
        }

        public static float GetNormalizedThumbLY(uint padNum = 0)
        {
            return InternalCalls.Input_GetNormalizedThumbLY(padNum);
        }

        public static float GetNormalizedThumbRX(uint padNum = 0)
        {
            return InternalCalls.Input_GetNormalizedThumbRX(padNum);
        }

        public static float GetNormalizedThumbRY(uint padNum = 0)
        {
            return InternalCalls.Input_GetNormalizedThumbRY(padNum);
        }
    }
}
